### PR TITLE
[JS] Modernize and automate code quality/formatting

### LINKED ISSUE #6657: Add prettier code formatter for JS, typescript

### LINKED ISSUE #6757: Revisit/revive/enforce `eslint` use

### PR DESCRIPTION
**Pull request by paulcam206**
# Related Issue

Fixes #6657
Fixes #6757 

# Description

The commits on this PR are structured such that each is self-contained:

* e6d74951455deaafd2919bcf0603e9ed24203253
  * Updated configuration for `eslint`
  * Install/configure `prettier`
  * `npm` scripts updated for `lint`
* d683d1cae0ce46563fe4ee2f83b9c031e0c983a5
  * Install/configure `lint-staged`, a tool that can run sets of tools against changes staged in `git`
* 1a54be7c0f25380d2cd16a10e0724b8f02096539
  * Install/configure `husky`, a client-side `git` hook manager
  * Wire `husky` up to our root `prepare` lifecycle script
* 5b55c77a156b88242ff0f8630e6410725fea8772
  * Run `eslint --fix` on the `adaptivecards` package - no manual edits
* 46bcce1bea81bf77b628b0a097d38f5ece90b721
  * Manual fixes for `eslint` issues. **This is probably the most important thing to review!**
* 0455e3b88595f4d94b57c53f3c31cdb7a2b47cdc
  * Run `prettier` on stuff in `adaptivecards` - no manual edits
* subsequent commits address PR feedback

With these changes, after this branch is merged, doing an `npm install` in `source/nodejs` will automatically install `husky`-managed `git` hooks. After a file under `source/nodejs` is modified and staged, attempting to commit it will cause `eslint` and `prettier` to run. Automatically-fixable issues detected by `eslint` will be fixed and any formatting changes that need to be made will also be done automatically. Any errors from `eslint` that *aren't* auto-fixable will yield an error and refuse the commit. This hook can be bypassed using the `--no-verify` `git` commandline argument.

I should note that I didn't apply any autofixing or formatting outside of the `adaptivecards` package for now, as each individual package will need its own dedicated time.

###### Microsoft Reviewers: [Open in CodeFlow](https://portal.fabricbot.ms/api/codeflow?pullrequest=https://github.com/microsoft/AdaptiveCards/pull/6815)

### CODE REVIEW

File: source/nodejs/.husky/pre-commit [hunk 1]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
+. "$(dirname "$0")/_/husky.sh"
```
Comment (licanhua) [inline source/nodejs/.husky/pre-commit:2]: Does it work on both windows and iOS? after run the commit command. will it still in / other than /source/nodejs?

File: source/nodejs/.husky/pre-commit [hunk 2]
```diff
@@ -0,0 +1,4 @@
...
+
+cd source/nodejs && npx lint-staged
```
Comment (licanhua) [inline source/nodejs/.husky/pre-commit:4]: how long does it take to finish the lint check?

File: source/nodejs/.lintstagedrc.mjs [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+      return eslint.isPathIgnored(file);
+    })
+  );
+  const filteredFiles = files.filter((_, i) => !isIgnored[i]);
+  return filteredFiles.join(" ");
+};
+
+export default {
+  "**/!(package-lock)*": async (files) => {
+    const filesToLint = await removeIgnoredFiles(micromatch(files, ["**/!(__)*.{ts,js}"]));
```
Comment (licanhua) [inline source/nodejs/.lintstagedrc.mjs:17]: Please format `tsx` too. I'm wondering if we have `.js`

File: source/nodejs/adaptivecards/src/__tests__/components/carousel.spec.ts [hunk 4]
```diff
@@ -18,28 +18,28 @@ describe('carousels', () => {
...
 
-    describe('when parsed from a valid object', () => {
+    describe("when parsed from a valid object", () => {
         const tinyTimer = 42;
 
         const carouselTheJson = {
-            "type": "Carousel",
-            "timer": tinyTimer,
-            "pages": [
+            type: "Carousel",
```
Comment (licanhua) [inline source/nodejs/adaptivecards/src/__tests__/components/carousel.spec.ts:25]: Personally I more like the style "type" as the key

File: source/nodejs/package.json [hunk 5]
```diff
@@ -6,7 +6,8 @@
 		"start:designer": "lerna run start --scope=adaptivecards-designer --stream",
 		"start:designer-app": "lerna run start --scope=adaptivecards-designer-app --stream",
 		"start:all": "lerna run --parallel watch && lerna run --parallel start",
-		"audit-all": "lerna-audit"
+		"audit-all": "lerna-audit",
+		"prepare": "cd ../.. && husky install source/nodejs/.husky"
```
Comment (licanhua) [inline source/nodejs/package.json:10]: It the format takes more than half an minutes or have a lot of outputs, I would like to opt in or at least put it into push hook.

File: source/nodejs/.husky/pre-commit [hunk 6]
```diff
@@ -0,0 +1,4 @@
+#!/bin/sh
```
Comment (licanhua) [inline source/nodejs/.husky/pre-commit:1]: remember to remove .ts/.tsx from .editconfig

File: source/nodejs/.eslintrc.js [hunk 7]
```diff
@@ -1,181 +1,89 @@
 module.exports = {
-    "root": true,
-    "extends": [
+    root: true,
+    extends: [
```
[the remark this answers is not in the document]
Comment (paulcam206) [reply source/nodejs/.eslintrc.js:3]: I found the `prettier` option to disable stripping quotes in this scenario :)

File: source/nodejs/.husky/pre-commit (hunk 2 above)
Comment (paulcam206) [reply source/nodejs/.husky/pre-commit:4]: it depends pretty highly on how much stuff is staged, but `eslint` runs on `source/nodejs/adaptivecards/src/*.ts` on my (not particularly beefy) machine in `3.8s`...

File: source/nodejs/package.json (hunk 5 above)
Comment (paulcam206) [reply source/nodejs/package.json:10]: formatting `source/nodejs/adaptivecards/**/*.ts` takes `6.45s` on my machine

Comment (paulcam206) [reply source/nodejs/package.json:10]: for what it's worth, linting and formatting is [recommended](https://git-scm.com/book/en/v2/Customizing-Git-Git-Hooks) in `pre-commit`

**Review by jwoo-msft — APPROVED** [review]

File: source/nodejs/.husky/pre-commit (hunk 1 above)
Comment (paulcam206) [reply source/nodejs/.husky/pre-commit:2]: works on Windows, iOS, and Linux. good point about running when there aren't files modified in `source/nodejs` - I added a little pre-filter here: 9971adb7386f0bc2fe8cbcbe50e9bb0057d4e36a

**Review by licanhua — APPROVED** [review]

**Pull request merged**